### ISSUE TITLE
dpo微调

### ISSUE THREAD

**Issue by yunpeili** [FunAudioLLM/CosyVoice#947]
目前的做法是instruct模型的基础上指令微调了一版，当作policy模型，然后用instruct+text去推理token，现在想用llm生成logits，可是会出现token和logits维度对不上所以算不了dpo loss；有没有做过的同学可以交流一下，或者官方有相关代码吗

**Comment by FlynnFlag** [comment]
我在用cosyvoice1做，我是自己重写了推理代码，让llm.inference输出每一个token和总词汇表的概率，再计算每个token的概率。LLM推理改成随机采样。其他不变

**Comment by yunpeili** [comment]
> 我在用cosyvoice1做，我是自己重写了推理代码，让llm.inference输出每一个token和总词汇表的概率，再计算每个token的概率。LLM推理改成随机采样。其他不变

是推理token的时候同时返回logp的意思吗？我也是cosyvoice1，之前也是改了llm.inference，但算logits不知道该不该让llm.forward

**Comment by FlynnFlag** [comment]
> > 我在用cosyvoice1做，我是自己重写了推理代码，让llm.inference输出每一个token和总词汇表的概率，再计算每个token的概率。LLM推理改成随机采样。其他不变
> 
> 是推理token的时候同时返回logp的意思吗？我也是cosyvoice1，之前也是改了llm.inference，但算logits不知道该不该让llm.forward

我是先改了inference 随机生成同样文本不同音频的数据，然后打分得到多个正负偏好，等造完再改llm的forward，直接输入造的数据的logp当reference model的部分来计算loss，还没做好，不知道可不可行

**Comment by FlynnFlag** [comment]
能跑通的
